File: core/PixelMapper.py
import numpy as np
import cv2
import json
import math

"""
Create an object for converting pixels to geographic coordinates,
using four points with known locations which form a quadrilteral in both planes
"""
class PixelMapper(object):
	def __init__(self, pixel_array, lonlat_array, lonlat_origin):
		#converting pixel and lonlat array to numpy array
		pixel_array = np.array(pixel_array)
		lonlat_array = np.array(lonlat_array)

		#identifying shape of arrays
		assert pixel_array.shape==(4,2), "Need (4,2) input array"
		assert lonlat_array.shape==(4,2), "Need (4,2) input array"

		#Calculates a perspective transform from four pairs of the corresponding points
		self.M = cv2.getPerspectiveTransform(np.float32(pixel_array),np.float32(lonlat_array))
		self.invM = cv2.getPerspectiveTransform(np.float32(lonlat_array),np.float32(pixel_array))
		
		#defining a few variables needed for conversions
		self.lonlat_origin = lonlat_origin
		self.lon_const = 40075000 * math.cos(lonlat_origin[0]*math.pi/180) / 360
		self.lat_const = 111320
        
	def pixel_to_lonlat(self, pixel):
		"""
		Convert a set of pixel coordinates to lon-lat coordinates
		"""
		#checking if pixel is of type np.ndarray and alters shape if not
		if type(pixel) != np.ndarray:
			pixel = np.array(pixel).reshape(1,2)
		assert pixel.shape[1]==2, "Need (N,2) input array" 
		#Joining the sequence of arrays along an existing axis
		pixel = np.concatenate([pixel, np.ones((pixel.shape[0],1))], axis=1)
		#calculating dot product of M and pixel.T, which is the lonlat coordinates
		lonlat = np.dot(self.M,pixel.T)

		#returning lonlat coordinates
		return (lonlat[:2,:]/lonlat[2,:]).T.tolist()
    
	def lonlat_to_pixel(self, lonlat):
		"""
		Convert a set of lon-lat coordinates to pixel coordinates
		"""
		#checking if lonlat is of type np.ndarray and alters shape if not
		if type(lonlat) != np.ndarray:
			lonlat = np.array(lonlat).reshape(1,2)
		assert lonlat.shape[1]==2, "Need (N,2) input array" 
		#Joining the sequence of arrays along an existing axis
		lonlat = np.concatenate([lonlat, np.ones((lonlat.shape[0],1))], axis=1)
		#calculating dot product of invM and lonlat.T, which are the pixel coordinates
		pixel = np.dot(self.invM,lonlat.T)

		#returning pixel coordinates
		return (pixel[:2,:]/pixel[2,:]).T.tolist()

	def lonlat_to_3D(self, lonlat):
		"""
		Convert a set of lon-lat coordinates to 3D coordinates
		"""
		lon_d = lonlat[0] - self.lonlat_origin[0]
		lat_d = lonlat[1] - self.lonlat_origin[1]

		lon_m = lon_d * self.lon_const
		lat_m = lat_d * self.lat_const

		#returning 3D coordinates
		return [lat_m, lon_m, 0]

	def _3D_to_lonlat(self, _coords3D):
		"""
		Convert a set of 3D coordinates to lon-lat coordinates
		"""
		lon_m = _coords3D[1]
		lat_m = _coords3D[0]
		
		lon_d = lon_m / self.lon_const
		lat_d = lat_m / self.lat_const
		
		lon_coord = lon_d + self.lonlat_origin[0]
		lat_coord = lat_d + self.lonlat_origin[1]
		
		#returning lon and lat coordinates
		return [lon_coord, lat_coord]

File: core_cpp/src/PixelMapper.cpp
// Copyright 2020 The Konect SpaSpect Authors. All Rights Reserved.
// This file is part of Konect SpaSpect technology.
//  ___            ___              _   
// / __|_ __  __ _/ __|_ __  ___ __| |_ 
// \__ \ '_ \/ _` \__ \ '_ \/ -_) _|  _|
// |___/ .__/\__,_|___/ .__/\___\__|\__|
//     |_|            |_|               
#include "PixelMapper.h"

// Convert a set of pixel coordinates to lon-lat coordinates
std::vector<Point2f*>* PixelMapper::pixel_to_lonlat(PixelMapperConfig &config, std::vector<Point2f*>& pixel_coordinates) {
	int N = pixel_coordinates.size();

	float pixel_matrix_transpose[3][N];
	for (int i = 0; i < N; ++i) {
		pixel_matrix_transpose[0][i] = static_cast<float>(pixel_coordinates[i]->x);
		pixel_matrix_transpose[1][i] = static_cast<float>(pixel_coordinates[i]->y);
		pixel_matrix_transpose[2][i] = 1.;
	}

	Mat pixel_matrix_transpose_converted = Mat(3, N, CV_32F, pixel_matrix_transpose);
	pixel_matrix_transpose_converted.convertTo(pixel_matrix_transpose_converted, CV_64F);
	Mat lonlat_matrix_transpose = config.M * pixel_matrix_transpose_converted;

	std::vector<Point2f*>* lonlat_coordinates = new std::vector<Point2f*>(N);

	for (int i = 0; i < N; ++i) {
		float scale_factor = lonlat_matrix_transpose.at<double>(i,2);
		Point2f* lonlat = new Point2f(0,0);
		lonlat->x = lonlat_matrix_transpose.at<double>(i,0) / scale_factor;
		lonlat->y = lonlat_matrix_transpose.at<double>(i,1) / scale_factor;
		(*lonlat_coordinates)[i] = lonlat;
	}

	return lonlat_coordinates;
}

// Convert a set of lon-lat coordinates to pixel coordinates
std::vector<Point2f*>* PixelMapper::lonlat_to_pixel(PixelMapperConfig &config, std::vector<Point2f*>& lonlat_coordinates) {
	int N = lonlat_coordinates.size();

	float lonlat_matrix_transpose[3][N];
	for (int i = 0; i < N; ++i) {
		lonlat_matrix_transpose[0][i] = static_cast<float>(lonlat_coordinates[i]->x);
		lonlat_matrix_transpose[1][i] = static_cast<float>(lonlat_coordinates[i]->y);
		lonlat_matrix_transpose[2][i] = 1.;
	}

	Mat lonlat_matrix_transpose_converted = Mat(3, N, CV_32F, lonlat_matrix_transpose);
	lonlat_matrix_transpose_converted.convertTo(lonlat_matrix_transpose_converted, CV_64F);
	Mat pixel_matrix_transpose = config.invM * lonlat_matrix_transpose_converted;

	std::vector<Point2f*>* pixel_coordinates = new std::vector<Point2f*>(N);

	for (int i = 0; i < N; ++i) {
		float scale_factor = pixel_matrix_transpose.at<double>(i,2);
		Point2f* pixel_coordinate = new Point2f(0,0);
		pixel_coordinate->x = pixel_matrix_transpose.at<double>(i,0) / scale_factor;
		pixel_coordinate->y = pixel_matrix_transpose.at<double>(i,1) / scale_factor;
		(*pixel_coordinates)[i] = pixel_coordinate;
	}

	return pixel_coordinates;
}

// Convert a set of lon-lat coordinates to 3D coordinates
std::vector<Point3f*>* PixelMapper::lonlat_to_3D(PixelMapperConfig &config, std::vector<Point2f*>& lonlat_coordinates) {
  int N = lonlat_coordinates.size();

  std::vector<Point3f*>* threeD_coordinates = new std::vector<Point3f*>(N);

  for (int i = 0; i < N; ++i) {
    	float lon_d = lonlat_coordinates[i]->x - config.lonlat_origin[i].x;
		float lat_d = lonlat_coordinates[i]->y - config.lonlat_origin[i].y;

		float lon_m = lon_d * config.lon_const;
		float lat_m = lat_d * config.lat_const;

		Point3f* threeD_coordinate = new Point3f(0,0,0);
		threeD_coordinate->x = lat_m;
		threeD_coordinate->y = lon_m;
		threeD_coordinate->z = 0;
		(*threeD_coordinates)[i] = threeD_coordinate;
	  }
  
	  return threeD_coordinates;
}

// Convert a set of 3D coordinates to lon-lat coordinates
std::vector<Point2f*>* PixelMapper::_3D_to_lonlat(PixelMapperConfig &config, std::vector<Point3f*>& threeD_coordinates) {
  int N = threeD_coordinates.size();
  float (*lonlat_coordinates_val)[2] = new float[N][2];

  std::vector<Point2f*>* lonlat_coordinates = new std::vector<Point2f*>(N);
  for (int i = 0; i < N; ++i) {
	  float lon_m = threeD_coordinates[i]->x;
	  float lat_m = threeD_coordinates[i]->y;

	  float lon_d = lon_m / config.lon_const;
	  float lat_d = lat_m / config.lat_const;

	  float lon_coord = lat_d + config.lonlat_origin->y;
	  float lat_coord = lon_d + config.lonlat_origin->x;

	  Point2f* lonlat_coordinate = new Point2f(0,0);
	  lonlat_coordinate->x = lat_coord;
	  lonlat_coordinate->y = lon_coord;
	  (*lonlat_coordinates)[i] = lonlat_coordinate;
  }
  return lonlat_coordinates;
}